pinecone/client.go

```diff
@@ -20,3 +41,357 @@ func (c *Client) IndexWithNamespace(host string, namespace string) (*IndexConnec
 	}
 	return idx, nil
 }
+
+func (c *Client) ListIndexes(ctx context.Context) ([]*Index, error) {
+	res, err := c.restClient.ListIndexes(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code: %d", res.StatusCode)
+	}
+
+	var indexList control.IndexList
+	err = json.NewDecoder(res.Body).Decode(&indexList)
+	if err != nil {
+		return nil, err
+	}
+
+	indexes := make([]*Index, len(*indexList.Indexes))
+	for i, idx := range *indexList.Indexes {
+		indexes[i] = toIndex(&idx)
+	}
+
+	return indexes, nil
+}
+
+type CreatePodIndexRequest struct {
+	Name             string
+	Dimension        int32
+	Metric           IndexMetric
+	Environment      string
+	PodType          string
+	Shards           int32
+	Replicas         int32
+	SourceCollection *string
+	MetadataConfig   *PodSpecMetadataConfig
+}
+
+func (req CreatePodIndexRequest) ReplicaCount() *int32 {
+	x := minOne(req.Replicas)
+	return &x
+}
+
+func (req CreatePodIndexRequest) ShardCount() *int32 {
+	x := minOne(req.Shards)
+	return &x
+}
+
+func (req CreatePodIndexRequest) TotalCount() *int {
+	x := int(*req.ReplicaCount() * *req.ShardCount())
+	return &x
+}
+
+func (c *Client) CreatePodIndex(ctx context.Context, in *CreatePodIndexRequest) (*Index, error) {
+	metric := control.IndexMetric(in.Metric)
+	req := control.CreateIndexRequest{
+		Name:      in.Name,
+		Dimension: in.Dimension,
+		Metric:    &metric,
+	}
+
+	//add the spec to req.
+	//because this is defined as an anon struct in the generated code, it must match exactly here.
+	req.Spec = control.CreateIndexRequest_Spec{
```

Comment:
I'm sort of not clear on why this is the case as opposed to the `Serverless` variety below. Primarily just not clear on how it's anon in the generated `CreateIndexRequest_Spec`.

Reply:
Yeah, it's unfortunate on the auto-gen code. On the Serverless autogen, it names the struct. Maybe it's used in more than one place? Either way, I consider it a bug in the generator. Luckily, because of these wrappers, folks using the SDK don't have to deal with this. 

As far as I know this is the only way in Go. I tried naming the `struct` shape locally but it considered it two different types.